core/src/main/java/io/questdb/std/DirectIntList.java

```diff
@@ -0,0 +1,185 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2023 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.std;
+
+import io.questdb.cairo.Reopenable;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.std.str.CharSink;
+
+import java.io.Closeable;
+
+public class DirectIntList implements Mutable, Closeable, Reopenable {
+
+    private static final Log LOG = LogFactory.getLog(DirectIntList.class);
+    private final long initialCapacity;
+    private final int memoryTag;
+    private long address;
+    private long capacity;
+    private long limit;
+    private long pos;
+    private long start;
+
+    public DirectIntList(long capacity, int memoryTag) {
+        this.memoryTag = memoryTag;
+        this.capacity = (capacity * Integer.BYTES);
+        this.address = Unsafe.malloc(this.capacity, memoryTag);
+        this.start = this.pos = address;
+        this.limit = pos + this.capacity;
+        this.initialCapacity = this.capacity;
+    }
+
+    public void add(int x) {
+        ensureCapacity();
+        assert pos < limit;
+        Unsafe.getUnsafe().putInt(pos, x);
+        pos += Integer.BYTES;
+    }
+
+    public final void add(DirectIntList that) {
+        long thatSize = that.pos - that.start;
+        if (limit - pos < thatSize) {
+            setCapacityBytes(this.capacity + thatSize - (limit - pos));
```

Comment:
this sizes the capacity precisely as needed to accommodate for `that` list.
this means the next `add()` is guaranteed to trigger another reallocation. Is this intentional?

Reply:
Yes, I think so. In many cases when we copy an off-heap list, we simply need the exact copy and we don't expand it after that.